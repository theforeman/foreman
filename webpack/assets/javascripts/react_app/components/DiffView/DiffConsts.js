export const SPLIT = 'split';
export const UNIFIED = 'unified';
