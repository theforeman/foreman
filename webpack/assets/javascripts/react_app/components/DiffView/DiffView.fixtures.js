import { noop } from '../../common/helpers';

export const diffMock = {
  oldText: 'hello friend',
  newText: 'hello there friend',
  viewType: 'split',
};

export const radioMock = {
  stateView: 'split',
  changeState: noop,
};

export const patchMock = {
  viewType: 'unified',
  patch: '\n---',
};
