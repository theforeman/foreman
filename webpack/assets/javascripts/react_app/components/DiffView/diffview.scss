#diff-radio-buttons {
  margin-bottom: 5px;
  float: right;

  .diff-button {
    text-transform: capitalize;
    width: 60px;
  }
}

#diff-table {
  border: 1px solid #d1d1d1;
  max-height: 400px;
  overflow: auto;
  clear: both;

  .diff-line {
    line-height: 1.7;
    font-family:
      'SFMono-Regular',
      Consolas,
      'Liberation Mono',
      Menlo,
      Courier,
      monospace;
  }

  .diff-gutter {
    border-left: 1px solid #e8e9ea;
  }

  .diff-gutter.diff-gutter-insert {
    background-color: #cdffd8;
  }

  .diff-hunk-header {
    border: 1px solid #e8e9ea;
  }

  .diff-code-insert {
    background-color: #e6ffed;

    mark.diff-code-edit {
      background-color: #acf2bd;
    }
  }

  .diff-code-delete {
    background-color: #ffeef0;

    mark.diff-code-edit {
      background-color: #fdb8c0;
    }
  }
}
