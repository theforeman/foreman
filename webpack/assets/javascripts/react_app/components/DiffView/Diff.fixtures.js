import { noop } from '../../common/helpers';

import { SPLIT, UNIFIED } from './DiffConsts.js';

export const patch = `
--- /etc/issue  2020-04-07 18:01:12.000000000 -0400
+++ /tmp/puppet-file20210218-2326-z6teom  2021-02-18 09:13:48.965177455 -0500
@@ -1,3 +1 @@
-\S
-Kernel \r on an \m
-
+fooo
\ No newline at end of file
`

export const diffMock = {
  oldText: 'hello friend',
  newText: 'hello there friend',
  viewType: SPLIT,
};

export const radioMock = {
  stateView: SPLIT,
  changeState: noop,
};

export const patchMock = {
  viewType: UNIFIED,
  patch,
};

export const fixtures = {
    'render DiffView w/oldText & newText': diffMock,
    'render DiffView w/Patch': patchMock,
  };

export const PF_SELECTED = 'pf-m-selected';

export const DIFF_SPLIT = 'diff-split';
export const DIFF_UNIFIED = 'diff-unified';
