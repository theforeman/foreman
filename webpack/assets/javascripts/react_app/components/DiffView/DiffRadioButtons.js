import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ButtonGroup, Button } from 'patternfly-react';
import { translate as __ } from '../../common/I18n';
import { SPLIT, UNIFIED } from './DiffConsts';

const btnClass = (stateView, btnView) =>
  classNames('diff-button', { active: stateView === btnView });

const radioButton = (stateView, btnView, changeState) => (
  <Button
    className={btnClass(stateView, btnView)}
    id={`${btnView}-btn`}
    onClick={changeState}
    bsStyle={stateView === btnView ? 'primary' : 'default'}
  >
    {btnView}
  </Button>
);

const DiffRadioButtons = ({ stateView, changeState }) => (
  <div id="diff-radio-buttons">
    <ButtonGroup>
      {radioButton(stateView, __('split'), () => {
        changeState(SPLIT);
      })}
      {radioButton(stateView, __('unified'), () => {
        changeState(UNIFIED);
      })}
    </ButtonGroup>
  </div>
);

DiffRadioButtons.propTypes = {
  stateView: PropTypes.string.isRequired,
  changeState: PropTypes.func.isRequired,
};

export default DiffRadioButtons;
