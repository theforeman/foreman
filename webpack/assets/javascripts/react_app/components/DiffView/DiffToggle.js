import React from 'react';
import PropTypes from 'prop-types';
import { ToggleGroup, ToggleGroupItem } from '@patternfly/react-core';
import { translate as __ } from '../../common/I18n';
import { SPLIT, UNIFIED } from './DiffConsts';

const DiffToggle = ({ stateView, changeState }) => {
  const handleItemClick = (isSelected, viewType) =>
    isSelected && changeState(viewType);

  return (
    <ToggleGroup id="diff-toggle-buttons" aria-label="editor-diff-buttons">
      <ToggleGroupItem
        key={`${SPLIT}-btn`}
        text={__('Split')}
        buttonId={`${SPLIT}-btn`}
        className="diff-button"
        isSelected={stateView === SPLIT}
        onChange={selected => handleItemClick(selected, SPLIT)}
      />
      <ToggleGroupItem
        key={`${UNIFIED}-btn`}
        text={__('Unified')}
        buttonId={`${UNIFIED}-btn`}
        className="diff-button"
        isSelected={stateView === UNIFIED}
        onChange={selected => handleItemClick(selected, UNIFIED)}
      />
    </ToggleGroup>
  );
};

DiffToggle.propTypes = {
  stateView: PropTypes.string.isRequired,
  changeState: PropTypes.func.isRequired,
};

export default DiffToggle;
