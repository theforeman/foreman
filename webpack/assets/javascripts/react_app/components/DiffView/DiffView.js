import React from 'react';
import PropTypes from 'prop-types';

import { parseDiff, Diff, markCharacterEdits } from 'react-diff-view';
import { formatLines, diffLines } from 'unidiff';
import 'react-diff-view/index.css';
import './diffview.scss';

const getDiff = (oldText, newText) => {
  const diffText = formatLines(diffLines(oldText, newText), { context: 3 });
  // these two lines are faked to mock git diff output
  const header = ['diff --git a/a b/b', 'index 0000000..1111111 100644'];
  return `${header.join('\n')}\n${diffText}`;
};

const DiffView = ({ oldText, newText, viewType, patch }) => {
  const markEdits = markCharacterEdits({
    threshold: 30,
    markLongDistanceDiff: true,
  });

  // old,new Text
  if (patch === '') {
    const gitDiff = getDiff(oldText, newText);
    const files = parseDiff(gitDiff);
    const hunk = files[0].hunks;

    return (
      hunk && <Diff hunks={hunk} markEdits={markEdits} viewType={viewType} />
    );
  }
  // Patch
  const files = parseDiff(
    patch
      .split('\n')
      .slice(1)
      .join('\n')
  );
  // eslint-disable-next-line react/prop-types
  const renderFile = ({ oldRevision, newRevision, type, hunks }) => (
    <Diff
      key={`${oldRevision}-${newRevision}`}
      viewType={viewType}
      diffType={type}
      hunks={hunks}
      markEdits={markEdits}
    />
  );

  return <div>{files.map(renderFile)}</div>;
};

DiffView.propTypes = {
  // None are required because only one can be used at a time: (old + new || patch)
  oldText: PropTypes.string,
  newText: PropTypes.string,
  viewType: PropTypes.string.isRequired,
  patch: PropTypes.string,
};

DiffView.defaultProps = {
  oldText: '',
  newText: '',
  patch: '',
};

export default DiffView;
