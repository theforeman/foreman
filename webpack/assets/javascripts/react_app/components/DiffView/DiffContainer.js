import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods } from '../../common/helpers';

import DiffView from './DiffView';
import DiffRadioButtons from './DiffRadioButtons';
import './diffview.scss';

class DiffContainer extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, ['changeState']);
    this.state = {
      viewType: 'split',
    };
  }

  changeState(viewType) {
    this.setState({ viewType });
  }

  render() {
    const { patch, oldText, newText } = this.props;
    const { viewType } = this.state;
    return (
      <div id="diff-container">
        <DiffRadioButtons changeState={this.changeState} stateView={viewType} />
        <div id="diff-table">
          <DiffView
            patch={patch}
            oldText={oldText}
            newText={newText}
            viewType={viewType}
          />
        </div>
      </div>
    );
  }
}

DiffContainer.propTypes = {
  oldText: PropTypes.string,
  newText: PropTypes.string,
  patch: PropTypes.string,
};

DiffContainer.defaultProps = {
  oldText: '',
  newText: '',
  patch: '',
};

export default DiffContainer;
