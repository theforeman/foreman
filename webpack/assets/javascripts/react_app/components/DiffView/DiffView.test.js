import { diffMock, patchMock } from './DiffView.fixtures';
import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';

import DiffView from './DiffView';

const fixtures = {
  'render DiffView w/oldText & newText': diffMock,
  'render DiffView w/Patch': patchMock,
};

describe('DiffView', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(DiffView, fixtures));
});
