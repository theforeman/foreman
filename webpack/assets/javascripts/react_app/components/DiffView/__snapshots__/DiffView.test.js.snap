// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffView rendering render DiffView w/Patch 1`] = `
<div>
  <Diff
    diffType="modify"
    generateAnchorID={[Function]}
    gutterAnchor={false}
    hideGutter={false}
    hunks={Array []}
    key="1111111-2222222"
    markEdits={[Function]}
    optimizeSelection={false}
    viewType="unified"
  />
</div>
`;

exports[`DiffView rendering render DiffView w/oldText & newText 1`] = `
<Diff
  diffType="modify"
  generateAnchorID={[Function]}
  gutterAnchor={false}
  hideGutter={false}
  hunks={
    Array [
      Object {
        "changes": Array [
          Object {
            "content": "hello friend",
            "isDelete": true,
            "lineNumber": 1,
            "type": "delete",
          },
          Object {
            "content": "hello there friend",
            "isInsert": true,
            "lineNumber": 1,
            "type": "insert",
          },
        ],
        "content": "@@ -1 +1 @@",
        "isPlain": false,
        "newLines": 1,
        "newStart": 1,
        "oldLines": 1,
        "oldStart": 1,
      },
    ]
  }
  markEdits={[Function]}
  optimizeSelection={false}
  viewType="split"
/>
`;
